Add tracer support

donaldong opened the pull request:
This allows users to add a tracer (for example, datadog's APM tracer) so the metrics/spans will be reported to the tracer.

### Test Plan
- [x] This integration works in a test project
- [x] Added test cases to cover the new behavior

donaldong commented:
update to use a block

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/8b1f4131cbeea079ef02569c13aebe012b21acd2..c059a15f45827685a1f345a3bd16b8f5ac92f052).

donaldong commented:
add doc

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/69cdcee3b477874833fd17e89aa8eb85c849cfdd..b173ff373a0b3254970d71b2c2d08982d9174646).

palasha commented on README.md:
```diff
@@ -109,6 +109,17 @@ Finished in 0.024 second
 Learn more: [Migrations](docs/migrations.md)
 
+### 6. Monitoring
+Redcord reports metrics to a tracer (for example, [Datadog APM](https://docs.datadoghq.com/tracing/setup/ruby/#manual-instrumentation)) if it is configured.
```
I would include something that makes the flexibility here more obvious. Eg. "The tracer any object with a `.trace(.....)` method". And then include datadog as an example.

palasha commented on lib/redcord.rb:
```diff
@@ -1,9 +1,36 @@
+
+  @@configuration_blks = T.let(
+    [],
+    T::Array[T.proc.params(arg0: T.untyped).void],
```
what if the config blocks have different lengths of arguments?

donaldong replied:
The blocks are only meant to be called by us here @ line 27

palasha commented on lib/redcord/prepared_redis.rb:
```diff
@@ -12,11 +12,16 @@ class Redcord::PreparedRedis < Redis
-      keys: [key],
-      argv: args.to_a.flatten,
-    ).to_i
+    Redcord::Base.trace(
```
I wonder if the public API methods could be auto instrumented instead of manually?

donaldong replied:
there's auto instrumentation for redis (if the supplied datadog.tracer is configured to do so).

or do you mean to do some runtime method generations to define those methods?

palasha replied:
> or do you mean to do some runtime method generations to define those methods?

Yea I guess I wasn't sure if there was already some automation in place. If there is, the tracing could be part of generating / instrumenting that code instead of manually.

palasha commented on lib/redcord/prepared_redis.rb:
```diff
@@ -12,11 +12,16 @@ class Redcord::PreparedRedis < Redis
-    ).to_i
+    Redcord::Base.trace(
+      'create_hash_returning_id',
+      resource: key,
```
nit: I'd consider having different hash arguments for different types, eg `key:` and `model:`, it might be easier for a subscriber to deal with that way?

donaldong replied:
good idea!

palasha commented on lib/redcord/tracer.rb:
```diff
@@ -0,0 +1,40 @@
+
+# typed: strict
+
+module Redcord::Tracer
```
I guess I find it a bit strange for this logic to be included into the various Redcord models, instead of being a "singleton" somewhere else that gets called into. Likely doesn't matter much but theoretically that would make it harder for someone to extend and mess with this in ways we wouldn't want.

donaldong replied:
currently, it's only mixed in Redcord::Base. it's using a `tracer` attribute that comes from the top-level configuration.

palasha commented on spec/tracer_spec.rb:
```diff
@@ -0,0 +1,40 @@
+  end
+
+  context 'when tracer is set' do
+    include Redcord::Tracer::ClassMethods
```
i had no idea that blocks could be used to scope includes/extends. I wonder how that works under the hood.

donaldong replied:
I think rspec generates a runtime class/module for each context block (so the classes, methods defined here wont conflict with each other)

palasha commented on lib/redcord/actions.rb:
```diff
@@ -35,13 +35,18 @@ def create!(args)
-      args = redis.hgetall(instance_key)
-      if args.empty?
-        raise Redcord::RecordNotFound, "Couldn't find #{name} with 'id'=#{id}"
+      Redcord::Base.trace(
```
what about tracing create! and find_by and the rest in this file?

donaldong replied:
that's a great idea! we should add tracing to both places. (actions.rb, and prepared_redis.rb (as named redis commands))

palasha replied:
My intent in suggesting automation (only if feasible and simple!) was to avoid missing Redcord API entrypoints like we might have here.

donaldong replied:
Yeah! Unfortunately seems dynamically inserting span would require some effort 😅  because model_name is not always determined prior to the method definition.

palasha reviewed (commented):
This is great.

I'd consider looking up OpenTracing to see if there's some obvious way to make this more aligned with their API, if its isn't already. That said I don't think its a big deal because this gives enough flexibility to build your own adapter.

donaldong commented:
address comments

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/ff06133b47449208bc94cd2a79478424fb157531..519db1467e1af3b67fcd4024e8ca492e2de43c98).

donaldong commented:
add span namespace

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/042e6605317fb3ae89f2d2d99ea6c317ff026024..c9577b54509e9f9205d50a17bb5c1c9e435327e6).

donaldong commented:
updated span name.

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/1ea8eb658545bfe53a6591f2c204e295edf211fe..4335e648efac6c72d90f89a105a3d91c742af962).

donaldong commented:
update api -- provide a block directly

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/c1cae1d45f2e4d4c94f8af422a6e14ad4aeed9a3..a075a0fc3db4f8661bdd3a000bb5946cf9ed7e3d).

katyho commented on lib/redcord.rb:
```diff
@@ -1,9 +1,36 @@
-module Redcord
-end
 
 require 'sorbet-runtime'
```
nit: can we keep these require statements together? Unless you separated them for a particular reason

donaldong replied:
yes, we used sorbet first in this file. Then we'd need to define module Redcord prior to other files

katyho commented on lib/redcord/actions.rb:
```diff
@@ -26,37 +26,62 @@ module ClassMethods
     sig { params(args: T::Hash[Symbol, T.untyped]).returns(Redcord::Relation) }
     def where(args)
-      Redcord::Relation.new(T.let(self, T.untyped)).where(args)
+      Redcord::Base.trace(
```
Where doesn't actually make any Redis call until one of these delegated methods is called: https://github.com/chanzuckerberg/redcord/blob/master/lib/redcord/relation.rb#L69

Can we also trace the delegated methods, or is it already doing that?

donaldong replied:
oooh, that's right! we should move this tracing to execute_query

donaldong commented:
add tracing to relation methods

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/b18a2dd912de74f42da0a7889468caf56595e571..3f8bd38e97b6b9638b9a7b46463ed86b8d771978).

donaldong commented:
remove tracing from count since it's only running a redis command

:mag_right: [Click to see changes since last update](https://github.com/chanzuckerberg/redcord/compare/0f7e561975d9d2d885f38736d47036185f333444..32f8e61c2f914fe26657d445a00fc07388b67a73).

katyho commented on spec/tracer_spec.rb:
```diff
@@ -0,0 +1,40 @@
+  context 'when tracer is set' do
+    include Redcord::Tracer::ClassMethods
+
+    let(:tracer) { proc { FakeTracer.new } }
```
Sorry maybe noob question, but I don't understand what this unit test is doing. Is tracer supposed to be used somewhere? e.g. `tracer.call.trace`?

donaldong replied:
Good question! when we call `trace`, it will use `tracer.call.trace` (tracer is mocked here, normally it'd set using `.tracer { ... }` )

katyho commented on spec/tracer_spec.rb:
```diff
@@ -0,0 +1,40 @@
+    it 'does not record traces' do
+      counter = 0
+      expect(
+        trace('test', model_name: 'test') { counter += 2 }
```
(Don't want to block on unit tests or my lack of understanding here) Intuitively I would expect this to be 0 then, if it's not supposed to record traces?

donaldong replied:
yeah its not recording traces, but we still need to run the block passed to this function.

katyho replied:
What in these tests is checking whether or not it's actually recording traces then?

donaldong replied:
ooh i see why its confusing now. We should expect the fake tracer class to receive the `trace` method.

the other test case has it set to nil so it won't receive anything.
